### ISSUE TITLE
OpenTelemetry Support

### ISSUE THREAD

**Issue by robbert229** [opentofu/opentofu#1519]
### OpenTofu Version

```shell
1.6.2
```

### Use Cases

# Problem

At scale it has become difficult to understand why certain invocations of OpenTofu are quick, and some take a long time. Its also not very easy to see the root causes of why an apply fails. Having a way of getting `tracing` information out of OpenTofu would allow for a much higher level of visibility into its operation, and empower platform maintainers (I maintain [tofutf](https://github.com/tofutf/tofutf)) to convey meaningful information to users.

Even for folks that don't use a platform, this still provides value. I know a number of teams that use the [Jenkins OpenTelemetry](https://plugins.jenkins.io/opentelemetry/) along side the [Ansible OpenTelemetry Plugin](https://docs.ansible.com/ansible/latest/collections/community/general/opentelemetry_callback.html). Having the option to do something similar, but with OpenTofu and the Jenkins OpenTelemetry plugin would provide a lot of value.

### Attempted Solutions

N/A

### Proposal

# Summary

My proposal is to instrument OpenTofu with the leading opensource standard for instrumentation, [OpenTelemetry](https://opentelemetry. Tracing is the only form of instrumentation that probably needs to be done, and it could be progressively adopted throughout the codebase without introducing much risk.

For example:

Adding an implementation of `github.com/opentofu/opentofu/internal/providers.Interface` that creates spans before calling an embedded implementation of the Interface would be one component of this, and would be a relatively small, and safe change.

# Candidates for instrumentation

* internal/providers.Interface
* internal/backend.Backend
* internal/getproviders
* internal/getmodules
* internal/httpclient
* internal/cloud
* plan creation
* plan application

# Final Thoughts

I am willing to invest some time here helping make this feature happen if its something that y'all think would be a good addition (Feel free to tell me that I am crazy though). Technically I think I can workaround the lack of opentelemetry integration by scripting up a program that scrapes logs via tflog, but I think that it would be better for the ecosystem if there was first class support here.

**Comment by robbert229** [comment]
Cool, I will hit him up via the opentofu slack and see if there is anything that I can help with regarding this.

**Comment by Yantrio** [comment]
I had a conversation with @robbert229 over the weekend. and Ill try and summarize it here.

Right now, I've been working on a branch to instrument the `plan` process using opentelemetry tracing. See this WIP PR here: https://github.com/opentofu/opentofu/pull/1517

This is a large amount of work due to the need to propagate the context.Context all the way down the callstacks.

I think this is something we can try and do piece by piece if people want it.

My initial concern is that whilst it's nice to add and a feature I would like to expand on over time, I would love to try and look at other quicker alternatives to try and solve people's problems.

I'll keep this issue open for now as I feel it is very valid, but I think maybe we could open up some other issues to try and solve the problems at hand there and discuss solutions to those.

For example, if someone is working on an automation platform (ie a TACOS), then they may want to pull out timing information about the run to highlight slow providers etc. This could be solved with telemetry, but it could also be solved by improving the story of machine readable logging and how we could produce that alongside human readable logs.

Another example is if someone wants to do some deep profiling of a slowdown, they may want to look into actually profiling instead of just using traces (However traces could get them 80% of the way there!)

Overall, I think this is a great idea, but if people have specific problem cases, maybe we could open an issue/discussion to look into the best solutions for those too! I don't think we should try and solve every problem with otel.

tldr; Hugely in favor of this issue, but it's going to take a while. Let's explore other solutions for visibility in the future though issues/discussions.
